packages/core/core-flows/src/order/workflows/create-return.ts

```diff
@@ -0,0 +1,305 @@
+} from "@medusajs/types"
+import {
+  createWorkflow,
+  transform,
+  WorkflowData,
+} from "@medusajs/workflows-sdk"
+import { createRemoteLinkStep, useRemoteQueryStep } from "../../common"
+import {
+  arrayDifference,
+  ContainerRegistrationKeys,
+  isDefined,
+  MathBN,
+  MedusaError,
+  Modules,
+} from "@medusajs/utils"
+import { updateOrderTaxLinesStep } from "../steps"
+import { createReturnStep } from "../steps/create-return"
+import { createFulfillmentWorkflow } from "../../fulfillment"
+
+function throwIfOrderIsCancelled({ order }: { order: OrderDTO }) {
+  // TODO: need work, check cancelled
+  if (false /*order.cancelled_at*/) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Order with id ${order.id} has been cancelled.`
+    )
+  }
+}
+
+function throwIfItemsDoesNotExistsInOrder({
+  order,
+  inputItems,
+}: {
+  order: Pick<OrderDTO, "id" | "items">
+  inputItems: OrderWorkflow.CreateOrderReturnWorkflowInput["items"]
+}) {
+  const orderItemIds = order.items?.map((i) => i.id) ?? []
+  const inputItemIds = inputItems.map((i) => i.id)
+  const diff = arrayDifference(inputItemIds, orderItemIds)
+
+  if (diff.length) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Items with ids ${diff.join(", ")} does not exist in order with id ${
+        order.id
+      }.`
+    )
+  }
+}
+
+function validateReturnReasons(
+  {
+    order_id,
+    inputItems,
+  }: {
+    order_id: string
+    inputItems: OrderWorkflow.CreateOrderReturnWorkflowInput["items"]
+  },
+  { container }
+) {
+  const reasonIds = inputItems.map((i) => i.reason_id).filter(Boolean)
+
+  if (!reasonIds.length) {
+    return
+  }
+
+  const remoteQuery = container.resolve(ContainerRegistrationKeys.REMOTE_QUERY)
+
+  const returnReasons = remoteQuery({
+    entry_point: "return_reasons",
+    fields: ["return_reason_children.*"],
+    variables: { id: [inputItems.map((item) => item.reason_id)] },
+  })
+
+  const reasons = returnReasons.map((r) => r.id)
+  const hasUnusableReasons = reasons.filter(
+    (reason) => reason.return_reason_children.length === 0
+  )
+  const hasUnExistingReasons = arrayDifference(reasonIds, reasons)
+
+  if (hasUnExistingReasons.length) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Return reason with id ${hasUnExistingReasons.join(
+        ", "
+      )} does not exists.`
+    )
+  }
+
+  if (hasUnusableReasons.length()) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Cannot apply return reason category with id ${hasUnusableReasons.join(
+        ", "
+      )} to order with id ${order_id}.`
+    )
+  }
+}
+
+function prepareShippingMethodData({
+  orderId,
+  inputShippingOption,
+  returnShippingOption,
+}: {
+  orderId: string
+  inputShippingOption: OrderWorkflow.CreateOrderReturnWorkflowInput["return_shipping"]
+  returnShippingOption: ShippingOptionDTO & {
+    calculated_price: { calculated_amount: number }
```

Comment:
suggestion: At some point, we should introduce a type `WithCalculatedPrice` that we can apply to all the places where we currently type it out explicitly. Think that will prove handy.